File: packages/sdks/@flagmind/core/src/openfeature.ts
import type { TombstoneClient } from './client.js';
import type { EvaluationContext } from './types.js';

// ─── OpenFeature interface definitions (inline — no peer dep required) ────────

export interface ResolutionDetails<T> {
  value: T;
  reason?: string;
  errorCode?: string;
  errorMessage?: string;
}

/** Minimal subset of the OpenFeature EvaluationContext contract. */
export interface OpenFeatureEvaluationContext {
  targetingKey?: string;
  [key: string]: unknown;
}

// ─── Reason mapping ───────────────────────────────────────────────────────────

const REASON_MAP: Record<string, string> = {
  OFF: 'DISABLED',
  FALLTHROUGH: 'DEFAULT',
  TARGET_MATCH: 'TARGETING_MATCH',
  RULE_MATCH: 'TARGETING_MATCH',
  PREREQUISITE_FAILED: 'DEFAULT',
  ERROR: 'ERROR',
};

function mapReason(flagMindReason: string): string {
  return REASON_MAP[flagMindReason] ?? 'UNKNOWN';
}

function buildEvaluationContext(
  context: OpenFeatureEvaluationContext | undefined
): EvaluationContext {
  const { targetingKey, ...rest } = context ?? {};
  const attrs: Record<string, string> = {};

  for (const [k, v] of Object.entries(rest)) {
    if (typeof v === 'string') {
      attrs[k] = v;
    } else if (v !== undefined && v !== null) {
      attrs[k] = String(v);
    }
  }

  return {
    userId: targetingKey ?? 'anonymous',
    ...(Object.keys(attrs).length > 0 ? { attrs } : {}),
  };
}

// ─── TombstoneProvider ─────────────────────────────────────────────────────────

/**
 * OpenFeature Provider for Tombstone.
 *
 * Usage:
 *   import { TombstoneClient } from '@tombstone/core';
 *   import { TombstoneProvider } from '@tombstone/core';
 *   import { OpenFeature } from '@openfeature/server-sdk';
 *
 *   const client = new TombstoneClient({ sdkKey: '...', environment: 'production', defaults: {} });
 *   const provider = new TombstoneProvider(client);
 *   await OpenFeature.setProviderAndWait(provider);
 */
export class TombstoneProvider {
  readonly metadata = { name: 'flagmind' };

  constructor(private readonly client: TombstoneClient) {}

  async initialize(): Promise<void> {
    if (!this.client.isConnected()) {
      await this.client.connect();
    }
  }

  async resolveBooleanEvaluation(
    flagKey: string,
    defaultValue: boolean,
    context?: OpenFeatureEvaluationContext
  ): Promise<ResolutionDetails<boolean>> {
    try {
      const ctx = buildEvaluationContext(context);
      const result = this.client.evaluate<boolean>(flagKey, ctx);
      return {
        value: typeof result.value === 'boolean' ? result.value : defaultValue,
        reason: mapReason(result.reason),
      };
    } catch (e: unknown) {
      return {
        value: defaultValue,
        reason: 'ERROR',
        errorCode: 'GENERAL',
        errorMessage: e instanceof Error ? e.message : String(e),
      };
    }
  }

  async resolveStringEvaluation(
    flagKey: string,
    defaultValue: string,
    context?: OpenFeatureEvaluationContext
  ): Promise<ResolutionDetails<string>> {
    try {
      const ctx = buildEvaluationContext(context);
      const result = this.client.evaluate<string>(flagKey, ctx);
      return {
        value: typeof result.value === 'string' ? result.value : defaultValue,
        reason: mapReason(result.reason),
      };
    } catch (e: unknown) {
      return {
        value: defaultValue,
        reason: 'ERROR',
        errorCode: 'GENERAL',
        errorMessage: e instanceof Error ? e.message : String(e),
      };
    }
  }

  async resolveNumberEvaluation(
    flagKey: string,
    defaultValue: number,
    context?: OpenFeatureEvaluationContext
  ): Promise<ResolutionDetails<number>> {
    try {
      const ctx = buildEvaluationContext(context);
      const result = this.client.evaluate<number>(flagKey, ctx);
      return {
        value: typeof result.value === 'number' ? result.value : defaultValue,
        reason: mapReason(result.reason),
      };
    } catch (e: unknown) {
      return {
        value: defaultValue,
        reason: 'ERROR',
        errorCode: 'GENERAL',
        errorMessage: e instanceof Error ? e.message : String(e),
      };
    }
  }

  async resolveObjectEvaluation<T extends object>(
    flagKey: string,
    defaultValue: T,
    context?: OpenFeatureEvaluationContext
  ): Promise<ResolutionDetails<T>> {
    try {
      const ctx = buildEvaluationContext(context);
      const result = this.client.evaluate<T>(flagKey, ctx);
      const isObject =
        typeof result.value === 'object' &&
        result.value !== null &&
        !Array.isArray(result.value);
      return {
        value: isObject ? result.value : defaultValue,
        reason: mapReason(result.reason),
      };
    } catch (e: unknown) {
      return {
        value: defaultValue,
        reason: 'ERROR',
        errorCode: 'GENERAL',
        errorMessage: e instanceof Error ? e.message : String(e),
      };
    }
  }
}

File: packages/sdks/tombstone-python-sdk/tombstone/openfeature.py
"""Tombstone OpenFeature provider — vendor-neutral feature flag evaluation."""
from __future__ import annotations

import logging
from dataclasses import dataclass, field
from typing import Any

from tombstone.client import TombstoneClient
from tombstone.types import EvaluationContext

logger = logging.getLogger(__name__)

# ─── OpenFeature types (inline — no peer dep required) ───────────────────────


@dataclass
class ResolutionDetails:
    """Mirrors the OpenFeature ResolutionDetails structure."""

    value: Any
    reason: str = "DEFAULT"
    error_code: str = ""
    error_message: str = ""


@dataclass
class OpenFeatureEvaluationContext:
    """Minimal subset of the OpenFeature EvaluationContext contract."""

    targeting_key: str = "anonymous"
    attributes: dict[str, str] = field(default_factory=dict)


# ─── Reason mapping ───────────────────────────────────────────────────────────

_REASON_MAP: dict[str, str] = {
    "OFF": "DISABLED",
    "FALLTHROUGH": "DEFAULT",
    "TARGET_MATCH": "TARGETING_MATCH",
    "RULE_MATCH": "TARGETING_MATCH",
    "PREREQUISITE_FAILED": "DEFAULT",
    "ERROR": "ERROR",
}


def _map_reason(flagmind_reason: str) -> str:
    return _REASON_MAP.get(flagmind_reason, "UNKNOWN")


def _build_context(context: OpenFeatureEvaluationContext | None) -> EvaluationContext:
    if context is None:
        return EvaluationContext(user_id="anonymous")
    return EvaluationContext(
        user_id=context.targeting_key or "anonymous",
        attrs=context.attributes or {},
    )


# ─── TombstoneProvider ─────────────────────────────────────────────────────────


class TombstoneProvider:
    """OpenFeature Provider for Tombstone.

    Usage::

        from tombstone import TombstoneClient
        from tombstone.openfeature import TombstoneProvider, OpenFeatureEvaluationContext

        client = TombstoneClient(sdk_key="...", environment="production")
        client.connect()
        provider = TombstoneProvider(client)

        ctx = OpenFeatureEvaluationContext(targeting_key="user-42")
        details = provider.resolve_boolean_details("payments.new-flow", False, ctx)
        print(details.value, details.reason)
    """

    def __init__(self, client: TombstoneClient) -> None:
        self._client = client

    @property
    def name(self) -> str:
        return "flagmind"

    # ── Boolean ──────────────────────────────────────────────────────────────

    def resolve_boolean_details(
        self,
        flag_key: str,
        default_value: bool,
        context: OpenFeatureEvaluationContext | None = None,
    ) -> ResolutionDetails:
        try:
            ctx = _build_context(context)
            result = self._client.evaluate(flag_key, ctx)
            value = result.value
            if not isinstance(value, bool):
                value = default_value
            return ResolutionDetails(value=value, reason=_map_reason(result.reason))
        except Exception as exc:  # noqa: BLE001
            logger.debug("Tombstone OpenFeature: resolve_boolean_details error: %s", exc)
            return ResolutionDetails(
                value=default_value,
                reason="ERROR",
                error_code="GENERAL",
                error_message=str(exc),
            )

    # ── String ───────────────────────────────────────────────────────────────

    def resolve_string_details(
        self,
        flag_key: str,
        default_value: str,
        context: OpenFeatureEvaluationContext | None = None,
    ) -> ResolutionDetails:
        try:
            ctx = _build_context(context)
            result = self._client.evaluate(flag_key, ctx)
            value = result.value
            if not isinstance(value, str):
                value = default_value
            return ResolutionDetails(value=value, reason=_map_reason(result.reason))
        except Exception as exc:  # noqa: BLE001
            logger.debug("Tombstone OpenFeature: resolve_string_details error: %s", exc)
            return ResolutionDetails(
                value=default_value,
                reason="ERROR",
                error_code="GENERAL",
                error_message=str(exc),
            )

    # ── Number ───────────────────────────────────────────────────────────────

    def resolve_number_details(
        self,
        flag_key: str,
        default_value: float | int,
        context: OpenFeatureEvaluationContext | None = None,
    ) -> ResolutionDetails:
        try:
            ctx = _build_context(context)
            result = self._client.evaluate(flag_key, ctx)
            value = result.value
            if not isinstance(value, (int, float)):
                value = default_value
            return ResolutionDetails(value=value, reason=_map_reason(result.reason))
        except Exception as exc:  # noqa: BLE001
            logger.debug("Tombstone OpenFeature: resolve_number_details error: %s", exc)
            return ResolutionDetails(
                value=default_value,
                reason="ERROR",
                error_code="GENERAL",
                error_message=str(exc),
            )

    # ── Object ───────────────────────────────────────────────────────────────

    def resolve_object_details(
        self,
        flag_key: str,
        default_value: object,
        context: OpenFeatureEvaluationContext | None = None,
    ) -> ResolutionDetails:
        try:
            ctx = _build_context(context)
            result = self._client.evaluate(flag_key, ctx)
            value = result.value
            if not isinstance(value, dict):
                value = default_value
            return ResolutionDetails(value=value, reason=_map_reason(result.reason))
        except Exception as exc:  # noqa: BLE001
            logger.debug("Tombstone OpenFeature: resolve_object_details error: %s", exc)
            return ResolutionDetails(
                value=default_value,
                reason="ERROR",
                error_code="GENERAL",
                error_message=str(exc),
            )
